ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/BrowserTabCloseHandler.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.ide.core;
+
+import static org.eclipse.che.ide.preferences.pages.general.IdeGeneralPreferencesPresenter.PREF_IDE_GENERAL_TAB_CLOSING;
+
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+import com.google.web.bindery.event.shared.EventBus;
+import org.eclipse.che.ide.api.WindowActionEvent;
+import org.eclipse.che.ide.api.WindowActionHandler;
+import org.eclipse.che.ide.api.editor.EditorAgent;
+import org.eclipse.che.ide.api.editor.EditorPartPresenter;
+import org.eclipse.che.ide.api.preferences.PreferencesManager;
+
+/** Process browser tab leave events. */
+public class BrowserTabCloseHandler implements WindowActionHandler {
+
+  private final Provider<EditorAgent> editorAgentProvider;
+  private final PreferencesManager preferencesManager;
+
+  /** Is used to not to block programmatic close of browser tab */
+  private static boolean closeImmediately = false;
+
+  /**
+   * If true given then no checks will be performed before IDE leave.
+   *
+   * <p>Typically should be invoked before automatic page leave/reloading.
+   */
+  public static void setCloseImmediately(boolean closeImmediately) {
+    BrowserTabCloseHandler.closeImmediately = closeImmediately;
+  }
+
+  @Inject
+  public BrowserTabCloseHandler(
+      EventBus eventBus,
+      Provider<EditorAgent> editorAgentProvider,
+      PreferencesManager preferencesManager) {
+    this.editorAgentProvider = editorAgentProvider;
+    this.preferencesManager = preferencesManager;
+
+    eventBus.addHandler(WindowActionEvent.TYPE, this);
+  }
+
+  @Override
+  public void onWindowClosing(WindowActionEvent event) {
+    if (!closeImmediately) {
+      if (!isEditorsClean()) {
+        event.setMessage("ask user");
```

Comment:
It looks like use will see the 'ask user' message.

Reply:
No, actually their won't. Modern browsers block custom messages, so now matters only whether it was null or a string. See `setMessage` description for more details.

Reply:
If browser starts taking into account messages then we will have to alter the message.
So, it is better to set the message if it is really displayed.

Reply:
It won't. At least the probability is too small (they deleted such ability on purpose).
But ok, I will add messages...